File: eclipse.py
"""
eclipse.py — the engine.

Local circumstances of the total solar eclipse of 12 August 2026, computed with
Skyfield and the JPL DE421 ephemeris.

Design rule: this module is *pure astronomy*. `compute_eclipse()` has no
knowledge of Streamlit, Folium, argparse or stdout, so the same function can be
imported by a web app, a REST API or a Discord bot without modification.
The CLI lives in `main()` and is the only place that prints.

    pip install skyfield skyfield-data rich geopy timezonefinder numpy
"""

from __future__ import annotations

import argparse
from dataclasses import dataclass
from datetime import datetime, timedelta, timezone
from math import acos, degrees, pi, radians, sqrt

import numpy as np
from skyfield.api import Loader, wgs84

# ── ephemeris: loaded ONCE at module level, never per call ─────────────────
_load = Loader("~/.skyfield-data", verbose=False)
TS = _load.timescale()
EPH = _load("de421.bsp")
EARTH, SUN, MOON = EPH["earth"], EPH["sun"], EPH["moon"]

# Physical constants (km)
R_SUN_KM = 696_000.0
R_MOON_KM = 1_737.4

# Global window of the event, UTC
EVENT_NAME = "Total Solar Eclipse — 12 August 2026"
GREATEST_ECLIPSE_UTC = datetime(2026, 8, 12, 17, 46, tzinfo=timezone.utc)
WINDOW_START = datetime(2026, 8, 12, 14, 30, tzinfo=timezone.utc)
WINDOW_END = datetime(2026, 8, 12, 21, 30, tzinfo=timezone.utc)
COARSE_STEP = timedelta(minutes=1)


@dataclass(frozen=True)
class City:
    name: str
    country: str
    lat: float
    lon: float


CITIES: list[City] = [
    City("Reykjavík", "Iceland", 64.1466, -21.9426),
    City("Akureyri", "Iceland", 65.6885, -18.1262),
    City("Nuuk", "Greenland", 64.1836, -51.7214),
    City("Longyearbyen", "Svalbard", 78.2232, 15.6469),
    City("Tórshavn", "Faroe Is.", 62.0079, -6.7908),
    City("A Coruña", "Spain", 43.3623, -8.4115),
    City("Oviedo", "Spain", 43.3619, -5.8494),
    City("Bilbao", "Spain", 43.2630, -2.9350),
    City("Zaragoza", "Spain", 41.6488, -0.8891),
    City("Valencia", "Spain", 39.4699, -0.3763),
    City("Palma", "Spain", 39.5696, 2.6502),
    City("Madrid", "Spain", 40.4168, -3.7038),
    City("Barcelona", "Spain", 41.3874, 2.1686),
    City("Lisbon", "Portugal", 38.7223, -9.1393),
    City("Paris", "France", 48.8566, 2.3522),
    City("London", "UK", 51.5072, -0.1276),
    City("Dublin", "Ireland", 53.3498, -6.2603),
    City("Berlin", "Germany", 52.5200, 13.4050),
    City("Moscow", "Russia", 55.7558, 37.6173),
    City("Rome", "Italy", 41.9028, 12.4964),
    City("New York", "USA", 40.7128, -74.0060),
]


# ── geometry helpers ───────────────────────────────────────────────────────
def _obscuration(sep: float, r_sun: float, r_moon: float) -> float:
    """Fraction of the solar disk area hidden by the lunar disk (all degrees)."""
    if sep >= r_sun + r_moon:
        return 0.0
    if sep <= abs(r_moon - r_sun):
        return 1.0 if r_moon >= r_sun else (r_moon / r_sun) ** 2
    d2, r1, r2 = sep * sep, r_sun, r_moon
    a1 = r1 * r1 * acos((d2 + r1 * r1 - r2 * r2) / (2 * sep * r1))
    a2 = r2 * r2 * acos((d2 + r2 * r2 - r1 * r1) / (2 * sep * r2))
    a3 = 0.5 * sqrt(
        max(0.0, (-sep + r1 + r2) * (sep + r1 - r2) * (sep - r1 + r2) * (sep + r1 + r2))
    )
    return (a1 + a2 - a3) / (pi * r1 * r1)


def _geometry(observer, t):
    """Topocentric Sun/Moon geometry at time `t`: (sep, r_sun, r_moon, alt)."""
    astro_sun = observer.at(t).observe(SUN).apparent()
    astro_moon = observer.at(t).observe(MOON).apparent()
    sep = astro_sun.separation_from(astro_moon).degrees
    d_sun = astro_sun.distance().km
    d_moon = astro_moon.distance().km
    r_sun = degrees(np.arcsin(R_SUN_KM / d_sun))
    r_moon = degrees(np.arcsin(R_MOON_KM / d_moon))
    alt = astro_sun.altaz()[0].degrees
    return float(sep), float(r_sun), float(r_moon), float(alt)


def _bisect(f, lo: float, hi: float, iterations: int = 26) -> float:
    """Root of `f` (unix seconds domain) bracketed by lo/hi."""
    f_lo = f(lo)
    for _ in range(iterations):
        mid = (lo + hi) / 2
        if (f(mid) >= 0) == (f_lo >= 0):
            lo = mid
        else:
            hi = mid
    return (lo + hi) / 2


def _dt(unix_seconds: float) -> datetime:
    return datetime.fromtimestamp(unix_seconds, tz=timezone.utc)


# ── the public, pure function ──────────────────────────────────────────────
def compute_eclipse(lat: float, lon: float) -> dict:
    """Local circumstances of the 2026-08-12 eclipse at (lat, lon).

    Returns a plain dict — no printing, no globals mutated, no side effects.
    Keys: visible, in_totality, max_obscuration, magnitude, totality_seconds,
    start, maximum, end, totality_start, totality_end, sun_altitude,
    sun_azimuth, profile.
    """
    observer = EARTH + wgs84.latlon(lat, lon)

    # coarse 1-minute scan for the moment of closest approach
    n = int((WINDOW_END - WINDOW_START) / COARSE_STEP) + 1
    times = [WINDOW_START + i * COARSE_STEP for i in range(n)]
    best_t, best_gap = times[0], float("inf")
    for moment in times:
        sep, r_s, r_m, _ = _geometry(observer, TS.from_datetime(moment))
        gap = sep - (r_s + r_m)
        if gap < best_gap:
            best_gap, best_t = gap, moment

    sep_at = lambda u: _geometry(observer, TS.from_datetime(_dt(u)))[0]  # noqa: E731
    lo = (best_t - COARSE_STEP).timestamp()
    hi = (best_t + COARSE_STEP).timestamp()
    for _ in range(40):  # ternary search on separation
        m1, m2 = lo + (hi - lo) / 3, hi - (hi - lo) / 3
        if sep_at(m1) < sep_at(m2):
            hi = m2
        else:
            lo = m1
    t_max = round((lo + hi) / 2)

    sep, r_sun, r_moon, alt = _geometry(observer, TS.from_datetime(_dt(t_max)))
    obsc = _obscuration(sep, r_sun, r_moon)
    result = {
        "lat": lat,
        "lon": lon,
        "any_contact": obsc > 0,
        "visible": obsc > 0 and alt > 0,
        "in_totality": False,
        "max_obscuration": obsc,
        "magnitude": max(0.0, (r_sun + r_moon - sep) / (2 * r_sun)),
        "totality_seconds": 0,
        "start": None,
        "maximum": _dt(t_max) if obsc > 0 else None,
        "end": None,
        "totality_start": None,
        "totality_end": None,
        "sun_altitude": alt,
        "sun_azimuth": 0.0,
        "profile": [],
    }
    if obsc <= 0:
        return result

    def touch(u: float) -> float:
        s, rs, rm, _ = _geometry(observer, TS.from_datetime(_dt(u)))
        return s - (rs + rm)

    def umbra(u: float) -> float:
        s, rs, rm, _ = _geometry(observer, TS.from_datetime(_dt(u)))
        return s - (rm - rs)

    def edge(f, direction: int, step: float, limit: int):
        u = float(t_max)
        for _ in range(limit):
            nxt = u + direction * step
            if f(nxt) > 0:
                return _bisect(f, u, nxt)
            u = nxt
        return None

    c1 = edge(touch, -1, 30.0, 400)
    c4 = edge(touch, +1, 30.0, 400)
    result["start"] = _dt(c1) if c1 else None
    result["end"] = _dt(c4) if c4 else None

    if r_moon > r_sun and umbra(t_max) < 0:
        t2 = edge(umbra, -1, 10.0, 200)
        t3 = edge(umbra, +1, 10.0, 200)
        if t2 and t3:
            result["in_totality"] = True
            result["totality_start"] = _dt(t2)
            result["totality_end"] = _dt(t3)
            result["totality_seconds"] = int(round(t3 - t2))

    if c1 and c4:
        result["profile"] = [
            (
                _dt(c1 + (c4 - c1) * i / 60),
                _obscuration(*_geometry(observer, TS.from_datetime(_dt(c1 + (c4 - c1) * i / 60)))[:3]),
            )
            for i in range(61)
        ]
    return result


def format_duration(seconds: int | float | None) -> str:
    """Human-readable duration: 214 -> '3m 34s'."""
    if not seconds or seconds <= 0:
        return "—"
    minutes, secs = divmod(int(round(seconds)), 60)
    return f"{secs}s" if minutes == 0 else f"{minutes}m {secs:02d}s"


def local_time(moment: datetime | None, lat: float, lon: float) -> str:
    """Format a UTC datetime in the local zone of (lat, lon)."""
    if moment is None:
        return "—"
    try:
        from timezonefinder import TimezoneFinder
        from zoneinfo import ZoneInfo

        name = TimezoneFinder().timezone_at(lat=lat, lng=lon)
        if name:
            return moment.astimezone(ZoneInfo(name)).strftime("%H:%M:%S %Z")
    except Exception:  # pragma: no cover - optional dependency
        pass
    return moment.strftime("%H:%M:%S UTC")


# ── CLI wrapper: the ONLY place with I/O ───────────────────────────────────
def main() -> None:
    parser = argparse.ArgumentParser(description=EVENT_NAME)
    parser.add_argument("lat", type=float, nargs="?", help="latitude, degrees north")
    parser.add_argument("lon", type=float, nargs="?", help="longitude, degrees east")
    parser.add_argument("--city", help="name of a featured city")
    args = parser.parse_args()

    if args.city:
        match = next((c for c in CITIES if c.name.lower() == args.city.lower()), None)
        if match is None:
            raise SystemExit(f"Unknown city: {args.city}")
        lat, lon, label = match.lat, match.lon, f"{match.name}, {match.country}"
    elif args.lat is not None and args.lon is not None:
        lat, lon, label = args.lat, args.lon, f"({args.lat:.3f}, {args.lon:.3f})"
    else:
        raise SystemExit("Give LAT LON or --city NAME")

    res = compute_eclipse(lat, lon)
    print(f"\n{EVENT_NAME}\n{label}  [{lat:.3f}, {lon:.3f}]")
    if not res["visible"]:
        print("  No eclipse visible — Sun below the horizon (or outside the shadow).")
        return
    if res["in_totality"]:
        print(f"  TOTAL ECLIPSE — {format_duration(res['totality_seconds'])} of totality")
        print(f"  totality: {local_time(res['totality_start'], lat, lon)}"
              f" → {local_time(res['totality_end'], lat, lon)}")
    else:
        print(f"  Partial eclipse — {res['max_obscuration'] * 100:.2f}% obscured")
    print(f"  first contact : {local_time(res['start'], lat, lon)}")
    print(f"  maximum       : {local_time(res['maximum'], lat, lon)}")
    print(f"  last contact  : {local_time(res['end'], lat, lon)}")
    print(f"  sun altitude  : {res['sun_altitude']:.1f}°  magnitude {res['magnitude']:.4f}\n")


if __name__ == "__main__":
    main()

File: src/lib/eclipse.ts
/*
 * eclipse.ts — the engine. Mirrors the public contract of eclipse.py:
 *
 *   computeEclipse(lat, lon) -> EclipseResult   (pure function, no UI, no I/O)
 *   CITIES                                       (featured markers)
 *   formatDuration(seconds) -> string
 *
 * Event: total solar eclipse of 12 August 2026.
 */

import { geometryAt, type Geometry } from './astro';

export const EVENT_NAME = 'Total Solar Eclipse — 12 August 2026';
/** Instant of greatest eclipse (UTC). */
export const GREATEST_ECLIPSE_UTC = Date.UTC(2026, 7, 12, 17, 46, 0);

/** Global search window for the event (UTC ms). */
const WINDOW_START = Date.UTC(2026, 7, 12, 14, 30, 0);
const WINDOW_END = Date.UTC(2026, 7, 12, 21, 30, 0);
const COARSE_STEP = 60_000; // 1 minute

export interface City {
  name: string;
  country: string;
  lat: number;
  lon: number;
}

export const CITIES: City[] = [
  { name: 'Reykjavík', country: 'Iceland', lat: 64.1466, lon: -21.9426 },
  { name: 'Akureyri', country: 'Iceland', lat: 65.6885, lon: -18.1262 },
  { name: 'Nuuk', country: 'Greenland', lat: 64.1836, lon: -51.7214 },
  { name: 'Longyearbyen', country: 'Svalbard', lat: 78.2232, lon: 15.6469 },
  { name: 'Tórshavn', country: 'Faroe Is.', lat: 62.0079, lon: -6.7908 },
  { name: 'A Coruña', country: 'Spain', lat: 43.3623, lon: -8.4115 },
  { name: 'Oviedo', country: 'Spain', lat: 43.3619, lon: -5.8494 },
  { name: 'Bilbao', country: 'Spain', lat: 43.263, lon: -2.935 },
  { name: 'Zaragoza', country: 'Spain', lat: 41.6488, lon: -0.8891 },
  { name: 'Valencia', country: 'Spain', lat: 39.4699, lon: -0.3763 },
  { name: 'Palma', country: 'Spain', lat: 39.5696, lon: 2.6502 },
  { name: 'Madrid', country: 'Spain', lat: 40.4168, lon: -3.7038 },
  { name: 'Barcelona', country: 'Spain', lat: 41.3874, lon: 2.1686 },
  { name: 'Lisbon', country: 'Portugal', lat: 38.7223, lon: -9.1393 },
  { name: 'Paris', country: 'France', lat: 48.8566, lon: 2.3522 },
  { name: 'London', country: 'UK', lat: 51.5072, lon: -0.1276 },
  { name: 'Dublin', country: 'Ireland', lat: 53.3498, lon: -6.2603 },
  { name: 'Berlin', country: 'Germany', lat: 52.52, lon: 13.405 },
  { name: 'Reykjahlíð', country: 'Iceland', lat: 65.6417, lon: -16.9167 },
  { name: 'Moscow', country: 'Russia', lat: 55.7558, lon: 37.6173 },
  { name: 'Rome', country: 'Italy', lat: 41.9028, lon: 12.4964 },
  { name: 'New York', country: 'USA', lat: 40.7128, lon: -74.006 },
];

/**
 * Approximate centre line of the path of totality (lat, lon), running
 * Siberia → Arctic Ocean → Greenland → Iceland → Atlantic → northern Spain
 * → Balearic Sea. Precomputed offline, baked in as a static constant so the
 * map costs nothing at runtime and works offline.
 */
export const CENTERLINE: [number, number][] = [
  [66.5, 158.0],
  [71.4, 148.5],
  [75.8, 137.0],
  [79.6, 121.5],
  [82.7, 99.0],
  [84.9, 68.0],
  [86.0, 28.0],
  [85.8, -12.0],
  [84.4, -41.0],
  [82.0, -58.0],
  [78.9, -66.0],
  [75.4, -68.0],
  [71.9, -64.5],
  [68.7, -57.5],
  [66.4, -47.0],
  [65.4, -35.0],
  [65.2, -25.2],
  [64.4, -18.0],
  [62.5, -11.5],
  [59.9, -7.0],
  [56.8, -4.0],
  [53.3, -2.0],
  [49.5, -1.0],
  [45.6, -1.5],
  [43.4, -3.4],
  [42.6, -1.4],
  [41.6, 0.7],
  [40.4, 2.6],
  [39.3, 4.2],
];

/** Rough umbral path edges (± half-width) for a shaded band. */
export function pathBand(halfWidthKm = 145): [number, number][] {
  const left: [number, number][] = [];
  const right: [number, number][] = [];
  for (let i = 0; i < CENTERLINE.length; i++) {
    const [lat, lon] = CENTERLINE[i];
    const [latA, lonA] = CENTERLINE[Math.max(0, i - 1)];
    const [latB, lonB] = CENTERLINE[Math.min(CENTERLINE.length - 1, i + 1)];
    const cosLat = Math.max(0.05, Math.cos((lat * Math.PI) / 180));
    let dx = (lonB - lonA) * cosLat;
    const dy = latB - latA;
    if (Math.abs(dx) > 180 * cosLat) dx -= Math.sign(dx) * 360 * cosLat;
    const len = Math.hypot(dx, dy) || 1;
    const nx = -dy / len;
    const ny = dx / len;
    const dDeg = halfWidthKm / 111.2;
    left.push([lat + ny * dDeg, lon + (nx * dDeg) / cosLat]);
    right.push([lat - ny * dDeg, lon - (nx * dDeg) / cosLat]);
  }
  return [...left, ...right.reverse()];
}

export interface EclipseResult {
  lat: number;
  lon: number;
  visible: boolean;
  anyContact: boolean;
  inTotality: boolean;
  maxObscuration: number;
  magnitude: number;
  totalitySeconds: number;
  startUTC: number | null;
  maxUTC: number | null;
  endUTC: number | null;
  totalityStartUTC: number | null;
  totalityEndUTC: number | null;
  sunAltitude: number;
  sunAzimuth: number;
  geometryAtMax: Geometry | null;
  /** obscuration samples through the event, for the sparkline */
  profile: { t: number; obs: number }[];
}

function bisect(
  f: (t: number) => number,
  lo: number,
  hi: number,
  iterations = 26,
): number {
  let a = lo;
  let b = hi;
  const fa = f(a);
  for (let i = 0; i < iterations; i++) {
    const mid = (a + b) / 2;
    if (Math.sign(f(mid)) === Math.sign(fa)) a = mid;
    else b = mid;
  }
  return (a + b) / 2;
}

/**
 * Pure function: local circumstances of the eclipse at (lat, lon).
 * No printing, no globals mutated, no framework imports. Import it from a web
 * app, a REST API, or a Discord bot without modification.
 */
export function computeEclipse(lat: number, lon: number): EclipseResult {
  // — coarse scan of the global event window —
  let bestT = WINDOW_START;
  let bestGap = Infinity;
  const profile: { t: number; obs: number }[] = [];

  for (let t = WINDOW_START; t <= WINDOW_END; t += COARSE_STEP) {
    const g = geometryAt(t, lat, lon);
    const gap = g.sep - (g.rSun + g.rMoon);
    if (gap < bestGap) {
      bestGap = gap;
      bestT = t;
    }
  }

  // — refine the maximum (ternary search on separation) —
  let lo = bestT - COARSE_STEP;
  let hi = bestT + COARSE_STEP;
  const sepAt = (t: number) => geometryAt(t, lat, lon).sep;
  for (let i = 0; i < 40; i++) {
    const m1 = lo + (hi - lo) / 3;
    const m2 = hi - (hi - lo) / 3;
    if (sepAt(m1) < sepAt(m2)) hi = m2;
    else lo = m1;
  }
  const maxT = Math.round((lo + hi) / 2 / 1000) * 1000;
  const gMax = geometryAt(maxT, lat, lon);

  const anyContact = gMax.obscuration > 0;
  const visible = anyContact && gMax.altitude > 0;

  const result: EclipseResult = {
    lat,
    lon,
    visible,
    anyContact,
    inTotality: false,
    maxObscuration: gMax.obscuration,
    magnitude: gMax.magnitude,
    totalitySeconds: 0,
    startUTC: null,
    maxUTC: anyContact ? maxT : null,
    endUTC: null,
    totalityStartUTC: null,
    totalityEndUTC: null,
    sunAltitude: gMax.altitude,
    sunAzimuth: gMax.azimuth,
    geometryAtMax: gMax,
    profile,
  };

  if (!anyContact) return result;

  // — contacts: sep − (rSun + rMoon) = 0 either side of maximum —
  const touch = (t: number) => {
    const g = geometryAt(t, lat, lon);
    return g.sep - (g.rSun + g.rMoon);
  };
  const searchEdge = (dir: 1 | -1) => {
    let t = maxT;
    for (let i = 0; i < 400; i++) {
      const next = t + dir * 30_000;
      if (next < WINDOW_START - 3600_000 || next > WINDOW_END + 3600_000) return null;
      if (touch(next) > 0) return bisect(touch, t, next);
      t = next;
    }
    return null;
  };
  result.startUTC = searchEdge(-1);
  result.endUTC = searchEdge(1);

  // — totality: sep − (rMoon − rSun) = 0 —
  const umbra = (t: number) => {
    const g = geometryAt(t, lat, lon);
    return g.sep - (g.rMoon - g.rSun);
  };
  if (gMax.rMoon > gMax.rSun && umbra(maxT) < 0) {
    result.inTotality = true;
    const edge = (dir: 1 | -1) => {
      let t = maxT;
      for (let i = 0; i < 200; i++) {
        const next = t + dir * 10_000;
        if (umbra(next) > 0) return bisect(umbra, t, next);
        t = next;
      }
      return null;
    };
    const a = edge(-1);
    const b = edge(1);
    if (a !== null && b !== null) {
      result.totalityStartUTC = a;
      result.totalityEndUTC = b;
      result.totalitySeconds = Math.round((b - a) / 1000);
    }
  }

  // — sample profile for the chart —
  const s = result.startUTC ?? maxT - 3600_000;
  const e = result.endUTC ?? maxT + 3600_000;
  const n = 60;
  for (let i = 0; i <= n; i++) {
    const t = s + ((e - s) * i) / n;
    profile.push({ t, obs: geometryAt(t, lat, lon).obscuration });
  }

  return result;
}

// ── caching layer (equivalent of @st.cache_data on the Python side) ─────────
const cache = new Map<string, EclipseResult>();

/** Round to ~100 m so tiny map movements reuse the cached computation. */
export function cachedEclipse(lat: number, lon: number): EclipseResult {
  const wrapped = ((((lon + 180) % 360) + 360) % 360) - 180;
  const rlat = Math.round(lat * 1000) / 1000;
  const rlon3 = Math.round(wrapped * 1000) / 1000;
  const key = `${rlat},${rlon3}`;
  const hit = cache.get(key);
  if (hit) return hit;
  const res = computeEclipse(rlat, rlon3);
  cache.set(key, res);
  return res;
}

export function formatDuration(seconds: number): string {
  if (!seconds || seconds <= 0) return '—';
  const m = Math.floor(seconds / 60);
  const s = Math.round(seconds % 60);
  if (m === 0) return `${s}s`;
  return `${m}m ${s.toString().padStart(2, '0')}s`;
}

export function fmtUTC(ms: number | null): string {
  if (ms === null) return '—';
  const d = new Date(ms);
  return (
    `${d.getUTCHours().toString().padStart(2, '0')}:` +
    `${d.getUTCMinutes().toString().padStart(2, '0')}:` +
    `${d.getUTCSeconds().toString().padStart(2, '0')} UTC`
  );
}

export function fmtLocal(ms: number | null): string {
  if (ms === null) return '—';
  return new Date(ms).toLocaleTimeString(undefined, {
    hour: '2-digit',
    minute: '2-digit',
    second: '2-digit',
  });
}

/** Longitude-based "sun clock" time — a decent stand-in for the local zone. */
export function fmtSolar(ms: number | null, lon: number): string {
  if (ms === null) return '—';
  const offset = Math.round(lon / 15) * 3600000;
  const d = new Date(ms + offset);
  const sign = offset >= 0 ? '+' : '−';
  return (
    `${d.getUTCHours().toString().padStart(2, '0')}:` +
    `${d.getUTCMinutes().toString().padStart(2, '0')}:` +
    `${d.getUTCSeconds().toString().padStart(2, '0')} (UTC${sign}${Math.abs(offset / 3600000)})`
  );
}

export function statusOf(res: EclipseResult): {
  label: string;
  tone: 'total' | 'deep' | 'partial' | 'slight' | 'none';
  color: string;
} {
  if (!res.anyContact || !res.visible)
    return { label: 'Not visible', tone: 'none', color: '#64748b' };
  if (res.inTotality) return { label: 'TOTAL', tone: 'total', color: '#22c55e' };
  const pct = res.maxObscuration * 100;
  if (pct >= 90) return { label: 'Deep partial', tone: 'deep', color: '#f59e0b' };
  if (pct >= 40) return { label: 'Partial', tone: 'partial', color: '#3b82f6' };
  return { label: 'Slight partial', tone: 'slight', color: '#6366f1' };
}
